Simplify handling of paths for input in BuildBamIndex.java

markjschreiber opened the pull request:
### Description

Changed to treat all input as a `Path` and use `IOUtil` to choose the path type so that paths with schemes handled by SPI providers are correctly handled as well as paths that have no scheme and are local files.

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [ ] Added or modified tests to cover changes and any new functionality
- [x] Edited the README / documentation (if applicable)
- [x] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

yfarjoun commented on src/main/java/picard/vcf/SplitVcfs.java:
```diff
@@ -73,16 +73,26 @@ public SplitVcfs() {
     @Override
     protected int doWork() {
-        IOUtil.assertFileIsReadable(INPUT);
+        Path inputPath;
```
final

yfarjoun commented on src/main/java/picard/vcf/SplitVcfs.java:
```diff
@@ -73,16 +73,26 @@ public SplitVcfs() {
+        } catch (IOException e) {
+            throw new RuntimeIOException(e);
+        }
+        IOUtil.assertPathsAreReadable(Stream.of(inputPath).collect(Collectors.toList()));
```
there's only one path here, no? why the stream and list?

markjschreiber replied:
There is no `assertPathIsReadable(Path p)` in `IOUtils` so I have to make a `List`, however I can use the more concise `Collections.singletonList(inputPath)`

yfarjoun commented on src/main/java/picard/vcf/processor/VariantIteratorProducer.java:
```diff
@@ -68,17 +60,30 @@
      */
     public static VariantIteratorProducer byHundredMegabaseChunksWithOnTheFlyFilteringByInterval(final List<File> vcfs, final IntervalList intervalList) {
-        return new Threadsafe(VcfFileSegmentGenerator.byWholeContigSubdividingWithWidth(ONE_HUNDRED_MILLION), vcfs, intervalList);
+        return new Threadsafe(VcfPathSegmentGenerator.byWholeContigSubdividingWithWidth(ONE_HUNDRED_MILLION), vcfs.stream().map(File::toPath).collect(Collectors.toList()), intervalList);
```
htsjdk.samtools.util.IOUtil#filesToPaths

yfarjoun reviewed (changes requested):
Thanks for this long overdue transition from file to path. 

Have you evaluated the option of using HtsPath? I know that we could change to that later on, but since this PR is opening many public APIs, I don't want to have to be in a situation of either maintaining or deprecating too many end points....

@cmnbroad perhaps you can opine since you are more aware of the readiness status of HtsPath (perhaps)....

markjschreiber commented:
> 

Thanks for the tip about `HtsPath`, that is more elegant. It seems `HtsPath.toPath()` already exists and returns an nio Path so I can use that. It works for paths with schemes (URIs) like `s3://foo/baa` and local files like `/path/to/file.bam`. Is there a need to override it?

cmnbroad commented:
Yes, HtsPath is what we use in GATK, since its always backed by a URI, and handles protocol schemes, query params, , absolute and relative local paths, url-encoding for things like embedded spaces, etc. We'd prefer to subclass it since it allows us to implement provider-specific functionality when loading file systems (we have needed to do that for GCS in GATK, and that can't reside in htsjdk since it would confer unwanted dependencies there). But I don't want to increase the tax too much here - if you can add the subclass, great, if not, just use HtsPath and I'll do the rest in a separate pass.

markjschreiber commented:
> Yes, HtsPath is what we use in GATK, since its always backed by a URI, and handles protocol schemes, query params, absolute and relative local paths, url-encoding for things like embedded spaces, etc. We'd prefer to subclass it since it allows us to implement provider-specific functionality when loading file systems (we have needed to do that for GCS in GATK, and that can't reside in htsjdk since it would confer unwanted dependencies there). But I don't want to increase the tax too much here - if you can add the subclass, great, if not, just use HtsPath and I'll do the rest in a separate pass.

Happy to subclass it but I can't see what is needed in the overloaded `toPath()` method. I had thought that NIO SPI implementations were not supposed to need special handling when loading other than being on the class path or the in the JDK extension directory?  The s3 nio spi that I have been developing seems to be automatically loaded by the JVM when it sees a URI with the `s3` scheme.

markjschreiber commented on src/main/java/picard/vcf/processor/VcfPathSegmentGenerator.java:
```diff
@@ -0,0 +1,189 @@
+ * Describes a mechanism for producing {@link VcfPathSegment}s from a VCF file path.
+ *
+ */
+public abstract class VcfPathSegmentGenerator {
```
True, although I think the duplication is with code from deprecated files which I assume would eventually be removed.

yfarjoun replied:
ah yes. I realized that eventually, and meant to remove my comment...sorry.

markjschreiber commented:
I have updated the code based on PR feedback including extending `HTSPath`. Let me know if you want additional changes.

yfarjoun reviewed (changes requested):
Thanks for the fast turnaround. I think that you were a little overzealous with the use of HtsPath. My understanding is that it should be used as a way of inputing commandline Strings into path-like objects, but that the code itself should use regular Path. @cmnbroad do yuo agree?

cmnbroad commented:
@yfarjoun @markjschreiber In my mind, the ultimate goal is to use the `HtsPath`-derived class as currency wherever possible, eliminating any use of, and interconversion between `File` or `String` (again, wherever possible); and converting just-in-time to `Path` at any API boundary that requires it (except htsjdk, as it becomes more ubiquitous there.) Part of the reason is related to another question @markjschreiber asked, which is why we need to subclass `HtsPath` in the first place. In GATK  we have had to put provider-specific functionality into the `HtsPath`-derived class for GCS and HDFS, to support things like file system loading, getting an input stream, etc. Ideally this shouldn't be required, but in practice it has been necessary, and it's so much easier if we have a central location to enable that kind of thing. Also,`HtsPath` already has some useful utilities such as `isCram`/'isBam`, etc., that I expect that to continue to evolve.

But ubiquitous `HtsPath` is a longer term goal - for now we don't have all the utility support we need to do that (i.e, assertIsReadable).  So we won't get there in this PR, but I think this is a big step forward in the right direction.

markjschreiber commented on src/main/java/picard/vcf/AccumulateVariantCallingMetrics.java:
```diff
@@ -75,7 +77,7 @@ protected int doWork() {
         final CollectVariantCallingMetrics.VariantCallingSummaryMetrics collapsedSummary = new CollectVariantCallingMetrics.VariantCallingSummaryMetrics();
 
-        List<Path> inputPaths = IOUtil.getPaths(INPUT);
+        final List<Path> inputPaths = INPUT.stream().map(PicardHtsPath::toPath).collect(Collectors.toList());
```
Done

markjschreiber commented:
> @yfarjoun @markjschreiber In my mind, the ultimate goal is to use the `HtsPath`-derived class as currency wherever possible, eliminating any use of, and interconversion between `File` or `String` (again, wherever possible); and converting just-in-time to `Path` at any API boundary that requires it (except htsjdk, as it becomes more ubiquitous there.) Part of the reason is related to another question @markjschreiber asked, which is why we need to subclass `HtsPath` in the first place. In GATK we have had to put provider-specific functionality into the `HtsPath`-derived class for GCS and HDFS, to support things like file system loading, getting an input stream, etc. Ideally this shouldn't be required, but in practice it has been necessary, and it's so much easier if we have a central location to enable that kind of thing. Also,`HtsPath` already has some useful utilities such as `isCram`/'isBam`, etc., that I expect that to continue to evolve.
> 
> But ubiquitous `HtsPath` is a longer term goal - for now we don't have all the utility support we need to do that (i.e, assertIsReadable). So we won't get there in this PR, but I think this is a big step forward in the right direction.

In my latest commit, I have reverted the use of `PicardHtsPath` for already existing public APIs but I have used them in new classes I have added to push further into a future where this becomes the standard Path type for Picard.

If there is interest I can continue to convert further Picard applications with the new pattern.

markjschreiber commented:
Hi @yfarjoun and @cmnbroad, I wasn't totally sure were you landed on the use of `Path` vs `PicardHtsPath` in the library but I think my last commit is in line with what you suggested. Can you review and let me know? 

If possible I'd like to merge soon and continue updating additional Picard applications to be able to use URIs for input.

yfarjoun commented on src/test/java/picard/vcf/AccumulateVariantCallingMetricsTest.java:
```diff
@@ -57,15 +61,15 @@
 
         final AccumulateVariantCallingMetrics program = new AccumulateVariantCallingMetrics();
-        program.INPUT = inputs;
+        program.INPUT = inputs.stream().map(PicardHtsPath::new).collect(Collectors.toList());
```
this should be a util method

markjschreiber replied:
Done

yfarjoun reviewed (approved):
looks good to me (barring one small request). 

@cmnbroad want to give a thumb regarding the use of path vs. htsjdkpath?

markjschreiber commented:
@yfarjoun, it looks like you need to approve running of the DSP AppSec check as well. As a first time contributor I can't do this.

cmnbroad commented on src/main/java/picard/vcf/SplitVcfs.java:
```diff
@@ -60,7 +65,7 @@
 
     @Argument(shortName = "D", doc = "The index sequence dictionary to use instead of the sequence dictionaries in the input files", optional = true)
-    public File SEQUENCE_DICTIONARY;
+    public String SEQUENCE_DICTIONARY;
```
Is there a reason not to use `PicardHtsPath` here ? Then I think you wouldn't need to add the try/catch block where it's used below to get the sequence dictionary.

markjschreiber replied:
Good catch

cmnbroad commented on src/test/java/picard/vcf/GatherVcfsTest.java:
```diff
@@ -61,18 +64,18 @@ public void setup() throws IOException {
         final File output = VcfTestUtils.createTemporaryIndexedFile("result", expectedOutput.getAbsolutePath().endsWith(".vcf") ? ".vcf" : ".vcf.gz");
 
-        inputFiles.forEach(f -> args.add("INPUT=" + f.getAbsolutePath()));
+        inputFiles.forEach(f -> args.add("INPUT=" + f.toPath().toUri()));
```
```suggestion
        inputFiles.forEach(f -> args.add("INPUT=" + f.getURI()));
```

markjschreiber replied:
Done

cmnbroad reviewed (approved):
Two very minor optional changes. Otherwise the HtsPath usage looks good to me.

markjschreiber commented:
@yfarjoun @cmnbroad , thanks for your approvals. I can't merge the PR as I don't have permission to run "DSP AppSec Trivy check" as a first time contributor. Could you run this check/ and or merge the PR? Thanks.

cmnbroad commented:
@markjschreiber Thanks so much for doing this!

markjschreiber commented:
Thanks for adding these changes!
